libvcx/src/aries/handlers/issuance/issuer/utils.rs

```diff
@@ -5,53 +5,72 @@ use crate::utils::error;
+
+                    // old style input such as {"address2":["101 Wilson Lane"]}
+                    serde_json::Value::Array(array_type) => {
+                        let attrib_value: &str = match array_type.get(0).and_then(serde_json::Value::as_str) {
+                            Some(x) => x,
+                            None => {
+                                warn!("Cannot encode attribute: {}", error::INVALID_ATTRIBUTES_STRUCTURE.message);
+                                return Err(VcxError::from_msg(VcxErrorKind::InvalidAttributesStructure, "Attribute value not found"));
+                            }
+                        };
+
+                        warn!("Old attribute format detected. See vcx_issuer_create_credential api for additional information.");
+                        attrib_value
+                    }
 
-    for (attr, attr_data) in attributes.iter_mut() {
-        let first_attr: &str = match attr_data {
-            // old style input such as {"address2":["101 Wilson Lane"]}
-            serde_json::Value::Array(array_type) => {
-                let attrib_value: &str = match array_type.get(0).and_then(serde_json::Value::as_str) {
-                    Some(x) => x,
-                    None => {
-                        warn!("Cannot encode attribute: {}", error::INVALID_ATTRIBUTES_STRUCTURE.message);
-                        return Err(VcxError::from_msg(VcxErrorKind::InvalidAttributesStructure, "Attribute value not found"));
+                    // anything else is an error
+                    _ => {
+                        warn!("Invalid Json for Attribute data");
+                        return Err(VcxError::from_msg(VcxErrorKind::InvalidJson, "Invalid Json for Attribute data"));
                     }
                 };
 
-                warn!("Old attribute format detected. See vcx_issuer_create_credential api for additional information.");
-                attrib_value
-            }
-
-            // new style input such as {"address2":"101 Wilson Lane"}
-            serde_json::Value::String(str_type) => str_type,
-            // anything else is an error
-            _ => {
-                warn!("Invalid Json for Attribute data");
-                return Err(VcxError::from_msg(VcxErrorKind::InvalidJson, "Invalid Json for Attribute data"));
+                let encoded = encode(&first_attr)?;
+                let attrib_values = json!({
+                    "raw": first_attr,
+                    "encoded": encoded
+                });
+
+                dictionary.insert(attr.to_string(), attrib_values);
+            };
+            serde_json::to_string_pretty(&dictionary)
+                .map_err(|err| {
+                    warn!("Invalid Json for Attribute data");
+                    VcxError::from_msg(VcxErrorKind::InvalidJson, format!("Invalid Json for Attribute data: {}", err))
+                })
+        }
+        Err(err) => { // TODO: Check error type
```

Comment:
This whole block of code if fairly hard to consume. If I get it right, `attributes` is something passed down by caller, and this code assumes it can be provided in various formats, right?
Originally there was expectation `attributes` is a map with `string` keys and some objects values.  And now this new block allows attributes to be Array of objects instead. 

If I get it right, the output format is always the same, right? This seems not to have impact on aries messaging, only allows additional vcx-specific input format when creating credential?

Reply:
> If I get it right, attributes is something passed down by caller, and this code assumes it can be provided in various formats, right?

Right.

> Originally there was expectation attributes is a map with string keys and some objects values. And now this new block allows attributes to be Array of objects instead.

Inaccurate - not object values, but string or array values, as you can see in the comments. The old format can probably be deprecated (which would simplify the code a little). We are using the "new" format (`{attr_name_1: attr_value_1, attr_name_2: attr_value_2, ...}`) everywhere. However, according to aries spec, format of the attrubutes is 
```
{
    "@type": "https://didcomm.org/issue-credential/1.0/credential-preview",
    "attributes": [
        {
            "name": "<attribute name>",
            "mime-type": "<type>",
            "value": "<value>"
        },
        // more attributes
    ]
}
```
This change is to support the aries format.

> If I get it right, the output format is always the same, right? This seems not to have impact on aries messaging, only allows additional vcx-specific input format when creating credential?

Correct, this change only allows new attributes format, the output is unchanged.

Reply:
Hmmm, sorry I must be missing something, but it seems to me right now, that modification in this PR does not alter anything on Aries level, isn't it? 

Looking at  definition `add_credential_preview_data` and related code it calls, vcx is already producing `credential-preview` with `attributes` in the format as you posted, eg.
```
"attributes": [ { "name": "<attribute name>", "mime-type": "<type>", "value": "<value>" },
```

It seems to me this PR extends AriesVCX credential API by an additional input format for credential values when creating credential, but messages which are subsequently produced in protocol are just the same.

Reply:
Have you read what I wrote? This is exactly what I am saying.

Reply:
Okay thank you for confirmation. I've noticed the changed not only impacts attribute encoding in `Credential` message, but also in `CredentialOffer` which got me a bit confused. It would probably had help me if PR had structured high-level description saying what has changed